cpp/dolfinx/io/XDMFFile.cpp

```diff
@@ -47,72 +53,161 @@ std::string get_hdf5_filename(std::string filename)
+#ifndef H5_HAVE_PARALLEL
+    if (mpi_io)
+    {
+      throw std::runtime_error(
+          "Cannot open file. HDF5 has not been compiled with support for MPI");
+    }
+#endif
+    _h5_id = HDF5Interface::open_file(_mpi_comm.comm(), hdf5_filename,
+                                      file_mode, mpi_io);
+    assert(_h5_id > 0);
+    LOG(INFO) << "Opened HDF5 file with id \"" << _h5_id << "\"";
+  }
+  else
+  {
+    // HDF handle be -1 to avoid closing a HDF file on destruction
+    _h5_id = -1;
+  }
+
+  if (_file_mode == "r")
+  {
+    // Load XML doc from file
+    pugi::xml_parse_result result = _xml_doc->load_file(_filename.c_str());
+    assert(result);
+
+    if (_xml_doc->child("Xdmf").empty())
+      throw std::runtime_error("Empty <Xdmf> root node.");
+
+    if (_xml_doc->child("Xdmf").child("Domain").empty())
+      throw std::runtime_error("Empty <Domain> node.");
+  }
+  else if (_file_mode == "w")
+  {
+    _xml_doc->reset();
+
+    // Add XDMF node and version attribute
+    _xml_doc->append_child(pugi::node_doctype)
+        .set_value("Xdmf SYSTEM \"Xdmf.dtd\" []");
+    pugi::xml_node xdmf_node = _xml_doc->append_child("Xdmf");
+    assert(xdmf_node);
+    xdmf_node.append_attribute("Version") = "3.0";
+    xdmf_node.append_attribute("xmlns:xi") = "http://www.w3.org/2001/XInclude";
 
-  // Check that encoding
-  if (_encoding == Encoding::ASCII and MPI::size(_mpi_comm.comm()) != 1)
+    pugi::xml_node domain_node = xdmf_node.append_child("Domain");
+    assert(domain_node);
+  }
+  else if (_file_mode == "a")
   {
-    throw std::runtime_error(
-        "Cannot write ASCII XDMF in parallel (use HDF5 encoding).");
+    if (boost::filesystem::exists(_filename))
+    {
+      // Load XML doc from file
+      pugi::xml_parse_result result = _xml_doc->load_file(_filename.c_str());
+      assert(result);
+
+      if (_xml_doc->child("Xdmf").empty())
+        throw std::runtime_error("Empty <Xdmf> root node.");
+
+      if (_xml_doc->child("Xdmf").child("Domain").empty())
+        throw std::runtime_error("Empty <Domain> node.");
+    }
+    else
+    {
+      _xml_doc->reset();
+
+      // Add XDMF node and version attribute
+      _xml_doc->append_child(pugi::node_doctype)
+          .set_value("Xdmf SYSTEM \"Xdmf.dtd\" []");
+      pugi::xml_node xdmf_node = _xml_doc->append_child("Xdmf");
+      assert(xdmf_node);
+      xdmf_node.append_attribute("Version") = "3.0";
+      xdmf_node.append_attribute("xmlns:xi")
+          = "http://www.w3.org/2001/XInclude";
+
+      pugi::xml_node domain_node = xdmf_node.append_child("Domain");
+      assert(domain_node);
+    }
   }
 }
 //-----------------------------------------------------------------------------
 XDMFFile::~XDMFFile() { close(); }
 //-----------------------------------------------------------------------------
 void XDMFFile::close()
 {
-  // Close the HDF5 file
-  _hdf5_file.reset();
+  if (_h5_id > 0)
```

Comment:
What about `_h5_id = 0`? Not clear in this function that case is properly handled.